zip-proportional-output-fee-mechanism-pofm.rst

```diff
@@ -0,0 +1,294 @@
+
+
+Requirements for gathering consensus
+------------------------------------
+
+Wallet developers SHOULD update the fees to the proposed formula under Specification section
+by Madars Virza and Kris Nuttycombe [#madars-1]_ 
+
+Where #inputs and #outputs also take into account transparent inputs and outputs. 
+Otherwise, the fee structure (if not otherwise changed) will preferentially encourage 
+usage of the transparent part of the protocol.
+
+The change to the conventional transaction fees must be undertaken soon
+as the Zcash network has been under heavy load with high-output transactions while 
+regular shielded transactions with 2 outputs are affected when relying on current 
+light wallet infrastructure.
+
+The following parties need to be part of the consensus:
+
+* The technical aspects of a changing conventional fee based on outputs 
+need to be evaluated.
+* A guarantee from mining groups is required to include the updated POFM 
+transactions in the next block.
+* Wallet developers need to update the software to use the new fee.
+* Zcash documentation and community outreach must be undertaken to
+make the change known.
+
+
+Requirements for adoption
+-------------------------
+
+The change to the conventional transaction fees should be undertaken soon
+as it gets difficult to gain consensus with the growth in the network
+of wallets, exchanges, miners, and third parties involved.
+
+The following parties need to be part of the consensus:
+
+* Support from mining groups is required to include the updated conventional
+  fee transactions in the next block.
+* Wallet developers need to provide a commitment to update the software to use
+  the new fee.
+* Zcash documentation and community outreach must be undertaken to make the
+  change known.
+* Changes to librustzcash crates to:
+  * expose minimum recommended fee for a given number of recipients
+  * implement default fee equation and corresponding unit tests
+  * integration of such mechanism to Transaction Builder APIs
+* Changes to the Zebra mempool's conventional fee calculation
+* adopt `librustzcash` changes to mobile SDK repos:
+  * ZcashLightClientKit
+  * zcash-light-client-ffi
+  * zcash-android-wallet-sdk
+
+
+Specification
+=============
+
+This specification defines three parameters that are used to calculate the
+conventional fee:
+
+=================== ============================
+Parameter           Units
+=================== ============================
+`base_fee`          zatoshis
+`marginal_fee`      zatoshis per input or output
+`grace_window_size` inputs or outputs
+=================== ============================
+
+Wallets implementing this specification SHOULD use a conventional fee
+calculated in zatoshis per the following formula::
+
+    base_fee + marginal_fee * max(0, inputs + outputs - grace_window_size)
+
+The parameters are set to the following values:
+* `base_fee = TODO`;
+* `marginal_fee = TODO`;
+* `grace_window_size = 4`.
```

Comment:
I advocate for `grace_window_size = 0`, otherwise this fee schedule incentivizes users to split up larger transactions into smaller transactions. I wrote up [a note about why it incentivizes this, the trade-offs, why I prefer 0, and why I'm against it philosophically](https://hackmd.io/@T_sFftPwS7uo67aa7yDfkA/SkUe8gdxs). A quick summary of why I advocate against incentivizing splitting up transactions:

The philosophical reason is we should aim for platform neutrality across usage patterns because Zcash is permissionless: anyone should be able to use it for any reason if it works for them and devs shouldn't interfere with that. Obviously, though, this has to be constrained by security and economic considerations, which is the case here, so this ZIP is addressing those economic & security concerns.

The practical motivations are that:
- UX: If users are incentivized to split up a single "need" into multiple transactions this adds a lot of usability headaches (ensuring all txns are correctly issued and confirmed, all counterparties verifying that, accommodating change outputs, etc…)
- Performance / Load: transactions carry a lot of overhead (relay, confirmation complexity, indexing, UI, scanning, etc…)

Reply:
In order to avoid an incentive for splitting transactions, we can either enforce the constraint that `grace_window_size * marginal_fee = base_fee`, or simplify as:
```
total_fee = marginal_fee * max(grace_window_size, inputs + outputs)
```